### ISSUE TITLE
test/train split in DRB

### ISSUE THREAD

**Issue by limnoliver** [USGS-R/delaware-model-prep#93]
@aappling-usgs and @jzwart [have been thinking about](https://docs.google.com/presentation/d/1zR0BV6aY5TmlASB4z3L-kNrJbKKvB7so0_MbyJILI0Q/edit?usp=sharing) how to properly make test/train splits for our modeling efforts in the DRB. This is a nuanced and important decision because:

- Temperature observations in the basin are skewed in time and space. Particularly with the onset of NGWOS in ~2017, there are many new sites later in the record. Some of these new sites capture dynamics that previous monitoring was not capturing (e.g., above/below reservoirs).
- Some key data are _only_ available in the NGWOS period. For example, Cannonsville and Pepacton reservoir temperature records are needed to calibrate GLM, but are only available starting in 2019. 
- Different models have different time coverages. PRMS-SNTemp is running from 1980-2020. The reanalyzed weather forecasts are only available from 2000-2019, etc. We need test/train splits that will cover each of these model windows.
- Test/train split decisions are important to make because we have several people working on different parts of the basin, different models, etc, and we want to be able to make comparisons across methods. We also want to be sure that we are leveraging the right data during training, and making test sets that cover gradients in the prediction challenge. 

We've already landed on a temporal test/train split ([see slide 3](https://docs.google.com/presentation/d/1zR0BV6aY5TmlASB4z3L-kNrJbKKvB7so0_MbyJILI0Q/edit#slide=id.gd0dea5ae0f_0_0)). But now we want to add some spatial holdouts. We're shooting for 4-10 sites, and want the following to be accomplished with those holdouts:

- [x] a headwater site
- [x] a mainstem site, preferably one of the flow target sites (Trenton, Montague?)
- [x] a paired above/below reservoir set of sites
- [x] a site in the Christina sub-basin that overlaps with a holdout that Xiaowei used (seg_id_nat: c('2007', '2012', '2013', '2014', and '2037'))
- [ ] Lordville
- [x] Holdout sites are representative of the basin across gradients in stream size (e.g., stream order or flow), in temperature dynamics, in urban vs more undisturbed settings, in proximity to reservoirs.

Some considerations/criteria:

- These holdout sites should have continuous records for a good chunk of time. I think a continuous year of data should be the minimum requirement. 
- When choosing a spatial holdout, we need to be sure there isn't information nearby that would be artificially creating good performance. e.g., a site immediately up/downstream of the holdout.
- I recommend _not_ holding out Cannonsville or Pepacton since observations there will likely be essential for making good predictions at Lordville. Candidate holdouts are around Beltzville and Neversink. 
- We want to balance good holdout sites that meet our above criteria with not losing too much training data. 
- One site could satisfy more than one of the above criteria.

Additionally, some questions this exercise should answer:

- Do the temporal holdouts create any spatial holdouts? e.g., are there sites that were only observed in the test period that could also be thought of as a spatial holdout? (essentially "free" spatial holdouts?)
- What's the proportion of total data that's being withheld? For space and time, and total space + time? Proportion of long term sites that are being withheld? 
- How are the spatial holdouts arranged geographically? Make sure we have good distribution across space. 
- Variations of [this figure](https://docs.google.com/presentation/d/1zR0BV6aY5TmlASB4z3L-kNrJbKKvB7so0_MbyJILI0Q/edit#slide=id.gcd2ba76313_0_36) should help us to visualize these decisions -- e.g., sorting sites by size, north to south, distance from reservoir, etc.

**Comment by wdwatkins** [comment]
What is the order of operations for temporal and spatial holdouts?  First we hold out the temporal set based on years defined in the slide (for all sites); then any additional spatial holdouts (beyond the 'free' ones) would be holding out data otherwise part of the training set?

**Comment by wdwatkins** [comment]
Should consider flow data as well, although not use different holdouts

**Comment by limnoliver** [comment]
Resources:

- should work from [this temperature](https://github.com/USGS-R/delaware-model-prep/blob/master/2_observations.yml#L80) and [this flow](https://github.com/USGS-R/delaware-model-prep/blob/master/2_observations.yml#L91) dataset. 
- proposed reservoir `seg_id_nat`s that have temperature observations - Neversink: up (1645) down (1638); Beltzville: up (1703) down (1697)
- This [internal mapper](http://delaware-basin-test-website.s3-website-us-west-2.amazonaws.com/) might be useful for poking around the basin.

**Comment by jzwart** [comment]
> I guess really I'm asking do spatial holdouts cover the entire period of record for those sites?

Yes, the spatial holdouts will withhold be the entire period of record (e.g. 1960-present) for testing, while the temporal holdouts will withhold 1980-1984, 2011-2015, & 2021 from all sites in the DRB for testing.

**Comment by wdwatkins** [comment]
Are there site ids somewhere for Cannonsville and Pepacton?  I assume these are not USGS sites, I don't see any matches in the site names.

**Comment by aappling-usgs** [comment]
@wdwatkins See [these slides](https://doimspp.sharepoint.com/:p:/r/sites/IIDDStaff/Shared%20Documents/Project%20-%20Modeling%20Streams/Reservoirs/DRB%20Reservoir%20IDs.pptx?d=w62ef5cc758154d58add281a7583b0a88&csf=1&web=1&e=eccElS) for many IDs to choose from.

**Comment by wdwatkins** [comment]
What constitutes a 'too-close' site for the spatial hold-outs?  

In spatial terms:
- Is any adjacent segment a concern, regardless of reach length (yes, if each reach is assigned a single temperature at all times?)
- Secondarily, should distance between reaches be considered as well (some reaches are quite short, so there could be little fish distance between reaches even if they are not adjacent.

In period of record terms:
 - After considering spatial relationships, are small numbers of discrete measurements in an adjacent reach a concern?  Or is it only a problem if there are nearby continuous sites?

**Comment by aappling-usgs** [comment]
I don't have, and I don't think we have, a clear idea of what "too close" will be, but we're hoping that seeing and discussing some maps with the proposed holdouts will help us figure that out. Do you think you could share something like:
* nearest reach with any observations and the mean and max obs per year at that site
* distance to nearest reach with a most-observed year having at least 200 daily observations

What's "too close" depends in part on how many better options are available - if all candidate holdout sites were within 10km of another hi-res observation site, for example, we'd need to accept that. Do you have any stats already that describe the options in these terms?

I think we mostly care about fish distance rather than the number of discretized reaches between two sites.

I'm not totally sure whether nearby sites with few observations can substantially improve a site or not. Close and highly observed sites are obviously influential. Sparsely observed sites feed less info into the overall loss function that drives model training, but it may still be that when the model is fruitfully overfitting to tailor predictions to each site (and it is), there's no reason for the model _not_ to learn a lot about a particular locality from just a few observations. So I do think that sparsely observed nearby sites are less of a concern than highly observed sites, and I wouldn't rule out a holdout site that had a sparsely observed nearby site unless there were clearly better options, but I also think a truly spatially isolated holdout site would be ideal if such sites exist and meet our other criteria.

**Comment by wdwatkins** [comment]
Yes, I have created most of those stats already, so I can base judgements on what is available.

Am I correct that the models estimate temperature at reach outlets, and account for reach length?  Does this mean it is reasonable to only consider outlet-to-outlet distances to assess closeness of observations (as opposed to looking at actual site locations, or considering adjacent reaches regardless of reach length)?

**Comment by jzwart** [comment]
Yes the model predicts temperature (and flow) at reach outlets, after taking into account processes that affect water temp and flow along the segment. I think outlet-to-outlet distances to assess closeness of observations is a good approach. I think you might also want 'bird distance' (or lat/lon of outlet) in addition to 'fish distance' of outlet-to-outlet - e.g. it would be good to know if there are two nearby segments by bird distance even if they have long fish distance (or aren't connected)

**Comment by wdwatkins** [comment]
What happens at confluences?  Are the mouths of the two upstream reaches treated separately in terms of their predicted temp/flow values, and only combined as inputs to the downstream reach?

**Comment by wdwatkins** [comment]
Here's some early figures.  I think rather than looking at some iteration of the closest reach, it might be better to look at total observations within a 'fish radius' in the updown matrix.  This should prevent heavily monitored sites being 'hidden' by a closer site with less data, and avoid a very distant, yet nearest, heavily monitored site having too much influence.  I can easily change the radius threshold being used.

I think this might be a good fit for a Jupyter notebook, so people can poke around these data if they want.

To start, here is the overall distribution of reach lengths, monitored or not:
![image](https://user-images.githubusercontent.com/13987349/116322443-f4319d00-a770-11eb-8b30-df4de66595fb.png)
Based on this, I think we should use at least a 10 km radius, maybe 15 or 20, to ensure the radius covers at least the median reach length up and down stream.  If there is a value that would be meaningful relative to how the models treat reach length, that might be helpful here.

Next, here are some of the sites we are biased towards holding out a priori, compared to a cdf of temperature observations within 10 or 20 km radius of monitored reaches (NA in all plots are two reaches with presumably continuous sites with the two highest period of record overlap with the temporal holdouts; 'Christina basin' is a headwater reach, so targeting that to hit two categories):

20 km fish radius:
![image](https://user-images.githubusercontent.com/13987349/116325442-044c7b00-a777-11eb-8c0a-cc8069333b68.png)

10 km fish radius, quite a bit different:
![image](https://user-images.githubusercontent.com/13987349/116324860-c569f580-a775-11eb-9c01-4dcb7c9a9d51.png)

A map of these sites in the basin:
![image](https://user-images.githubusercontent.com/13987349/116324077-40caa780-a774-11eb-999f-9fadf7d3b152.png)

**Comment by aappling-usgs** [comment]
Cool way of thinking about this, David!

Where is Montague in the 10km figure?

Velocity is slower in headwaters, so it might be appropriate to use more like the 10km metric for headwaters but 20km for the mainstem. But then, we're constrained in our options for the mainstem. How many other highly-observed sites are available along the mainstem, and where?

I like the idea of holding out that southernmost gray site because the topography seems to be quite different there, so it'd be a strong test of the models. Is water flowing south there? 

How many observations are available, and in what years, at the sites you've highlighted?

**Comment by aappling-usgs** [comment]
> What happens at confluences? Are the mouths of the two upstream reaches treated separately in terms of their predicted temp/flow values, and only combined as inputs to the downstream reach? If there were only one predicted value for the two upstream reaches, that would impact the mapping of upstream reaches.

I remember checking that the mouths of two upstream reaches are treated separately from each other and the downstream reach. I think [this is the code chunk](https://github.com/USGS-R/delaware-model-prep/blob/master/1_network/src/geo_fabric_functions.R#L66-L81) that further splits reaches that break that rule, though there were only one or two reaches that did, and when they did, I think it was because one upstream reach and the downstream reach were unified rather than because both upstream reaches shared data. Also note that the further splitting I did is only evidenced in `subseg_id`, not in `seg_id_nat`, which come from the unmodified NHGF.

**Comment by jzwart** [comment]
Nice! these are useful figures. 

I agree that having the southern most grey site would be good to have as a spatial holdout.  So potential spatial holdout sites could be: 

1. Lordville (important for management) 
2. Christina (headwater)
3. Southern most NA site 
4. Beltzville Reservoir ( paired upstream / downstream?) 
5. Montague (mainstem and important for management)

**Comment by wdwatkins** [comment]
> Where is Montague in the 10km figure?
It is at zero, underneath the Christina reach.  I didn't see a quick solution to dealing with overlaps in that plot.

Looks like the southern gray site is the Maurice River, it does flow south, I didn't notice that.  There is a reservoir on it, Union Lake, 2 reaches downstream of the one on the map.  Part of it is a National Wild and Scenic River.

Here are other mainstem reaches with at least 400 temperature observations:
![image](https://user-images.githubusercontent.com/13987349/116493260-316e5b80-a853-11eb-8aa5-925629f8d150.png)

Plotted as number of temp observations vs observations within a 20km fish radius:
![image](https://user-images.githubusercontent.com/13987349/116493600-e86ad700-a853-11eb-8aca-3f6496b31e78.png)
64_1: Trenton/Morrisville
134_1:  Above Lackawaxen R at Barryville
144_1: Callicoon
333_1: Florence/Edgely/Bristol (only one downstream of Trenton, might be good for a further downstream reach)

I'll look at covered time ranges tomorrow.

**Comment by wdwatkins** [comment]
Continuing with mainstem reaches, here are periods of record for reaches with at least 400 observations, ordered north to south:
![image](https://user-images.githubusercontent.com/13987349/116614363-b35f9280-a8ee-11eb-9f32-127589bf47a2.png)

For the select gages shown earlier (candidate reservoirs, Trenton/Montague/Lordville, Maurice River and other):
![image](https://user-images.githubusercontent.com/13987349/116614614-ff123c00-a8ee-11eb-85bf-8d2a68759cbf.png)

Labeled map of these seg_ids for reference:
![image](https://user-images.githubusercontent.com/13987349/116614675-15b89300-a8ef-11eb-8824-9fbfe16cb69f.png)

**Comment by wdwatkins** [comment]
A few thoughts, based on what I've looked at and being fairly ignorant of the management and regulatory aspects here:

Lower mainstem sites: Trenton has by far the longest period of record.  It is also fairly well-situated in terms of minimal nearby measurements within 20km fish radius (see scatter plot).  It would make for a very thorough test site for evaluating the model performance for several decades, but would also wipe out a lot of training data.  A combination of Montague and Florence/Edgely/Bristol (333) being held out would give a sparser coverage of a similar time range.

Upper mainstem sites: 
If there are not a-priori reasons to choose Lordville, I'd also suggest considering Callicoon (144_1) over Lordville (139), since it has a similar length of record and has fewer data within 20km fish radius, while still being in the same part of the basin.

Above/below:
Beltzville seems a better choice than Neversink based on nearby observations.

**Comment by jzwart** [comment]
These are nice summaries @wdwatkins . Also, the Trenton segment has _a lot_ of observations. I'm a bit worried about removing too much training data with that site - could you calculate proportion of training data removed if sites were to be withheld? 

Thinking about when the observations occur, do we want a spatial holdout sites to have more observations during the later decades? Or doesn't that matter as much? I'm thinking about segment `2136_1` that has more obs during the early 1980's to withhold. 

Good point about Callicoon segment. The Lordville segment should have more reservoir influence so might be a better test than the Callicon segment

**Comment by wdwatkins** [comment]
Yeah, I was wondering that about older data too.  Is there greater priority on this modeling performing well under current conditions than performing well through time?  And is there any indication from results so far on how performance varies through time?

I'll make a plot of the fraction of total data at each site next, then look into representativeness compared to discharge stats.  What would be a useful measure of temperature dynamics?  Overall minimum and maximum temperature?  Seems tricky to come up with something robust given the mix of continuous and discrete data

**Comment by wdwatkins** [comment]
Considering just time holdouts, we're at 81/18 train/test for temperature, and 87/13 for flow (~3.5m total observed reach-days for flow, ~.49m for temp).  In a more conventional non-spatial problem I think that would give us reasonable leeway to withhold more test data if we are only splitting into two groups.  I'm not sure how relevant that is here though.

Trenton is indeed the largest chunk of temperature data, at a little over 4% of the total:
![image](https://user-images.githubusercontent.com/13987349/116645105-ef641900-a929-11eb-8750-89e416ebbce0.png)

Considering only the additional data loss from a spatial holdout, the distribution looks fairly similar:
![image](https://user-images.githubusercontent.com/13987349/116645406-a496d100-a92a-11eb-9451-7194ca0cedaa.png)

**Comment by limnoliver** [comment]
These are great, information-rich figures! One thing to consider is that we're not doing any modeling prior to 1980. I'm not sure how much that affects these calculations/considerations, but certainly inflates the contribution from those few sites that have data since the 60's.

**Comment by wdwatkins** [comment]
Yeah, that is a good point.  So by not doing any modeling, I assume that is related to lack of process model data before that?  Is the pre-1980 data useful for anything here then (a sort of secondary test set, if we can't train without the process model?)

**Comment by jzwart** [comment]
Mostly related to a lack of meteorological drivers before 1980. I don't see the current projects using observations earlier than 1980 so I don't think it should be included in the train/test split decisions.

**Comment by wdwatkins** [comment]
I updated the percent data figures from yesterday to account for that ☝️ 

For now I'm using the Beltzville above/below segments, Lordville, Trenton, the Christina headwater, and the Maurice River as holdouts to check their distribution in catchment attributes and affect on data percentages.  

These segments would hold out an additional 9% of temperature data and 2.8% of flow, bringing splits to 69/31 for temp and 73/27 for flow.

**Distribution of catchment attributes**
I added some discharge metrics to the catchment attributes.  Visually, I think we are covering most variables reasonably well.  The only question mark is `cov_type` (land cover).  This is the metdata definition:

> Land cover type designation.  Cover type is first designated for each cell in the input raster map of land cover data. This is done by reclassifying the input land cover raster data set. As a quality assurance measure, any cells whose elevation exceeds timberline (set to 11,500 ft) and have shrub or tree designations are reset to grass (code 2). cov_type is set to 0 for a feature if the cover type cells within the feature aremore than 90% bare. If move than 20% of the feature is covered by tree cells, then cov_type is set to 3. For features that are unassigned a cov_type value by the preceding, a code of 2 is assigned to any feature with more than 20% shrub cover. Features that are still unassigned but with a combination of shrub and tree cover that exceeds 35% are then assigned a value. cov_type code 3 is assigned to features within this set with more tree than shrub area, with the remainder being assigned a value of 2. Any features still not designated with a cov_type are assigned according to the more commonly occuring (per-cell) cov_type within the feature. The final step is to re-assign the cov_type for any feature whose hru_elev value exceeds timberline (11,500 feet) and whose cov_type value is 2 or 3 to 1.

@jsadler2 is there more description of these values somewhere?  I assume the '1' values aren't really above timberline here.
![image](https://user-images.githubusercontent.com/13987349/116738605-88854500-a9a7-11eb-93e9-1ff3aa3c00ae.png)

I could add some measure of temperature dynamics to this plot.

**Comment by limnoliver** [comment]
I think capturing some range of urban areas is important in the basin - so we could potentially target a site at least 20% impervious? 

![image](https://user-images.githubusercontent.com/15788176/116742822-8d54f380-a9bd-11eb-9e76-46c9bcc184f5.png)

**Comment by limnoliver** [comment]
And I think a 70/30 train/test split is about right!

**Comment by jzwart** [comment]
Nice, yeah I agree that the 70/30 train/test split is good and agree that getting a segment with a slightly higher percent impervious would be good (although the spatial holdouts match the distribution fairly well right now). I think this is getting close to deciding on what segments to withhold for the spatial holdout. At first I was a little hesitant about withholding Trenton site but the overall train/test split seems about right even with the Trenton site. Here's a summary of candidate segments so far with rankings (withhold, maybe withhold, unsure), let me know if this is a correct summary: 

**Definitely withhold**
1. Christina headwater
2. Beltzville upstream 
3. Beltzville downstream 

**Maybe withhold (pick at least 3 of the following segments to withhold)**
1. Lordville 
2. Callicoon
2. Trenton 
3. Montague 
4. Maurice River
5. Segment with 0.2 fraction impervious landcover 

I think Maurice river should probably be  selected since it is far southern part of DRB. Then I think we should pick one segment from Lordville or Callicoon, and Trenton or Montague. And then maybe one more segment that is ~0.2 fraction impervious landcover if we don't represent that with the other segments withheld. That would put us at 6-7 segments withheld and close to 70/30 split for train/test.

**Comment by wdwatkins** [comment]
I think I would nominate subseg 904_1 as an additional reach cover a highly impervious area.

It is the best observed of those with >0.2 impervious, and has minimal observations within 20 km fish radius:
![image](https://user-images.githubusercontent.com/13987349/116748917-fc7a1a00-a9b4-11eb-950a-6beedce77546.png)

It isn't very near other segments we're targeting (compare to similar plots above):
![image](https://user-images.githubusercontent.com/13987349/116749021-2df2e580-a9b5-11eb-9fa1-9246fc97b7a6.png)

It has a few recent years of continuous data:
![image](https://user-images.githubusercontent.com/13987349/116749140-57ac0c80-a9b5-11eb-8250-13d5dc0adfaa.png)

376 and 332 are both mainstem reaches, and are already listed on the mainstem reaches plot above.  332 is close to Trenton so is ruled out by that.  376 is the farthest downstream mainstem reach.  It could be acceptable, though doesn't have very long record.

**Comment by jsadler2** [comment]
> @jsadler2 is there more description of these values somewhere? I assume the '1' values aren't really above timberline here.

No. Not that I'm aware of. And yeah. I think you are right. I don't think any of these basins are above the timber line, so it's strange that any would be marked as "1" based on that definition.

**Comment by wdwatkins** [comment]
Here's the mainstem reaches in terms of temperature observations within 20km fish radius, with our prospects colored:
![image](https://user-images.githubusercontent.com/13987349/116755484-567fdd00-a9bf-11eb-98b4-3f77ea25a4f9.png)
Trenton and Callicoon would be the best based on this.

**Comment by wdwatkins** [comment]
Trenton also provides more geographic diversity than Montague.  If we did pick Montague over Trenton, we might also pick 376 to still include a lower mainstem site.  Both have much less data than Trenton.

Lordville and Callicoon aren't too far apart, and have similar amounts of data (Callicoon has a bit more).  The main differentiator data-wise is observations within 20km fish radius.

**Comment by wdwatkins** [comment]
> I think capturing some range of urban areas is important in the basin - so we could potentially target a site at least 20% impervious?

Should there be any bias towards a smaller stream for this, that might be more impacted by urban runoff?  The Schuylkill is pretty big, although I'm not sure there are well-observed alternatives.  @jsadler2  which/how many HRUs is HRU percent impervious including for each reach?

**Comment by wdwatkins** [comment]
With Beltzville, Christina headwater, Maurice River, Trenton, and Callicoon, there is good coverage in terms of distance below reservoirs.  Negative values do not have any reservoirs upstream.
![image](https://user-images.githubusercontent.com/13987349/116762021-09efce00-a9ce-11eb-89b8-282ea717eb1b.png)

**Comment by wdwatkins** [comment]
Based on the sites in `2_observations/out/drb_filtered_sites.rds`, the fractions of heldout sites from different sources looks very reasonable:
![image](https://user-images.githubusercontent.com/13987349/116911299-887d8300-abfb-11eb-9526-a70714527d7a.png)

I eliminated some duplicates of sites with the same site_id, source, and site_type that only had different original_source.  It didn't make a large difference.

**Comment by jzwart** [comment]
>Should there be any bias towards a smaller stream for this, that might be more impacted by urban runoff? 

My guess is that this is going to be hard to find for this dataset and/or we'd have to be modeling lower-order streams than PRMS geospatial fabric. 

So should the final spatial holdouts be: 

1. Beltzville upstream
2. Beltzillve downstream
3. Christina headwater
4. Maurice River
5. Trenton
6. Callicoon 
7. Schuylkill at PHL

**Comment by limnoliver** [comment]
It seems like there should be more than ~40 reaches that are downstream of a reservoir?
![reservoir_reach_map](https://user-images.githubusercontent.com/15788176/116919590-1c107d00-ac17-11eb-9017-8133fb29f112.png)

**Comment by wdwatkins** [comment]
The dot plot above is only showing monitored reaches.  I can check that map ☝️  against a plot of monitored reaches downstream of reservoirs though.

**Comment by limnoliver** [comment]
Ah yeah, that explains at least some of it!

**Comment by wdwatkins** [comment]
Actually it is only monitored reaches with at least 400 observations (i.e. that would be candidates for spatial holdouts).  That is a good point though—the dot plots are all just showing the distribution of sites with enough observations to hold out.  I suppose really we should also be showing unmonitored reaches to be sure we are representing the full distribution of reaches, not just monitored reaches!  That should be a quick fix.

Here is a map of reservoir polygons (blue) and reaches (>400 observed days) colored by distance below reservoir.  I think this checks out, although it did miss the two reaches that fully contain a reservoir but are also headwaters (so looking upstream in the distance matrix missed them).
![image](https://user-images.githubusercontent.com/13987349/116928184-94743f80-ac11-11eb-8a13-110d2cba29c1.png)

**Comment by wdwatkins** [comment]
Ok, here is the grid of dotplots including all reaches (holdout = FALSE includes unmonitored reaches:
![image](https://user-images.githubusercontent.com/13987349/116932496-30ed1080-ac17-11eb-8b3c-92012efa047d.png)
There is still reasonable coverage of most variables, though Jensen-Haise coefficient (`jh_coef`) and snow area threshold (`snarea_thresh`) aren't the best.  Note there are some reaches without catchment attributes, so those aren't shown here.  ggplot doesn't seem to do a great job scaling the axes everytime, I think some very tall bins are cut off at the top.

**Comment by limnoliver** [comment]
Just clarifying: in the above map with reservoirs, the black reaches are those with <400 obs? I think this looks right, then.

**Comment by wdwatkins** [comment]
Yes

**Comment by limnoliver** [comment]
David, can you do a final tally of test/train split for flow and temp from the temporal holdouts + sites that Jake proposed?

**Comment by wdwatkins** [comment]
For these final sites:
![image](https://user-images.githubusercontent.com/13987349/117193000-3a04eb80-ad97-11eb-92bf-09a7b07d0a9e.png)

68/32 train/test for temperature, 72/28 for flow
Available data ordered N-> S with holdouts identified:
![image](https://user-images.githubusercontent.com/13987349/117192867-0fb32e00-ad97-11eb-8232-f5ff6fd1ee1e.png)

I suppose one could argue these sites are a bit skewed towards the lower parts of the basin.  Would it be worth adding a far northern headwater?

**Comment by limnoliver** [comment]
My vote would be to remove Maurice since we were selecting it as a "southern" representative, but now we know we have lots of other south-basin sites. We could then pick another upper basin site.

**Comment by wdwatkins** [comment]
The Maurice flows south into Delaware bay, so does not drain into the Delaware main stem (not very visible on that map ☝️ ).   Is it ok to lose coverage of that part of the basin from the spatial holdouts?

**Comment by jzwart** [comment]
I think I'd be OK keeping Maurice, there seems to be more data coverage spatially in southern reaches and maybe we can afford to take out one of those segments for testing? 

If we were to drop Maurice for a more northern segment, I think the northern segment should be non-mainstem and not downstream of a reservoir

**Comment by wdwatkins** [comment]
What would be our floor for % training data?  We could keep Maurice and add an upper basin site.  Filtering down to sites upstream of all reservoirs, > 400 oberved days and <1200 observations within 20 km fish radius, we would be removing between .1% and .7% additional data

**Comment by jzwart** [comment]
Yeah maybe adding another northern segment in addition to Maurice would be fine.

**Comment by aappling-usgs** [comment]
These maps and figures and discussion have all been great. I agree with keeping Maurice held out and adding another northern, reservoir-free headwater (doesn't need to be a long-term site, just a few good years of obs). And then I think we're good to go.

**Comment by limnoliver** [comment]
Great.  Thanks all. David, I think the remaining tasks to put a bow on this is to - 1) select one more northern holdout per Alison's description, 2) create a new "test" column in the [observation file](https://github.com/USGS-R/delaware-model-prep/blob/master/2_observations.yml#L76) with logicals indicating which observations should be held out for both temporal and spatial holdouts, and 3) create a paragraph that summarizes the spatial (with seg_id_nat numbers) and temporal (with date ranges) decisions. 

Thank you all for the mental and coding energy that went into this!

**Comment by wdwatkins** [comment]
I've added Willowemoc Creek (1455).  It has a few years of observations, and only a few hundred observed days within 20km fish radius.

Here are the same plots as above with the new set of 8 sites.  I also noticed I was using calendar years instead of water years above, that is now fixed 😬 .  That caused a small reduction in the training fraction—with the new site, it is now 66.2/33.8 for temperature, 71.4/28.6 for flow.

All of the other above plots should be reproducible with the final test set with minimal effort if needed from my PR.  I will add the summary paragraph there as well.

![image](https://user-images.githubusercontent.com/13987349/117492328-4f5c5000-af26-11eb-92c2-b0dd8afcd4f2.png)
![image](https://user-images.githubusercontent.com/13987349/117492528-90ecfb00-af26-11eb-9545-24c940cf2c8d.png)
